syntax = "proto2";

message Person {
  required string name = 1;
  required int32 id = 2;
  optional string email = 3;

  enum PhoneType {
    MOBILE = 0;
    HOME = 1;
    WORK = 2;
  }

  message PhoneNumber {
    required string number = 1;
    optional PhoneType type = 2 [default = HOME];
  }

  repeated PhoneNumber phone = 4;
}

message AddressBook {
  repeated Person contacts = 1;
}

message SearchResult {
  repeated Person results = 1;
}

message AddAddressReply {
  optional string message = 1;
}

message SearchRequest {
  optional string search_string = 1;
}

service AddressBookService {
    rpc AddAddress (Person) returns (AddAddressReply) {};
    rpc AllAddresses (SearchRequest) returns (AddressBook) {};
    rpc SearchAddresses (SearchRequest) returns (SearchResult) {};
}